securesystemslib/interface.py

```diff
@@ -120,237 +104,194 @@ def get_password(prompt='Password: ', confirm=False):
+  """Decryption password helper.
 
-  # Encrypt the private key if 'password' is set.
-  if len(password):
-    private = securesystemslib.keys.create_rsa_encrypted_pem(private, password)
+  - Fail if 'password' is passed and 'prompt' is True (precedence unclear)
+  - Return None on empty pw on prompt (suggests desire to not decrypt)
 
-  else:
-    logger.debug('An empty password was given.  Not encrypting the private key.')
+  """
+  securesystemslib.formats.BOOLEAN_SCHEMA.check_match(prompt)
 
-  # If the parent directory of filepath does not exist,
-  # create it (and all its parent directories, if necessary).
-  securesystemslib.util.ensure_parent_dir(filepath)
+  # We don't want to decide which takes precedence so we fail
+  if password is not None and prompt:
+    raise ValueError("passing 'password' and 'prompt=True' is not allowed")
 
-  # Write the public key (i.e., 'public', which is in PEM format) to
-  # '<filepath>.pub'.  (1) Create a temporary file, (2) write the contents of
-  # the public key, and (3) move to final destination.
-  file_object = tempfile.TemporaryFile()
-  file_object.write(public.encode('utf-8'))
-  # The temporary file is closed after the final move.
-  securesystemslib.util.persist_temp_file(file_object, filepath + '.pub')
+  # Prompt user for password
+  if prompt:
+    password = get_password("enter password to decrypt private key file "
+        "'" + TERM_RED + str(path) + TERM_RESET + "' "
+        "(leave empty if key not encrypted): '", confirm=False)
 
-  # Write the private key in encrypted PEM format to '<filepath>'.
-  # Unlike the public key file, the private key does not have a file
-  # extension.
-  file_object = tempfile.TemporaryFile()
-  file_object.write(private.encode('utf-8'))
-  securesystemslib.util.persist_temp_file(file_object, filepath)
+    # Treat empty password as no password. A user on the prompt can only
+    # indicate the desire to not decrypt by entering no password.
+    if not len(password):
+      return None
 
-  return filepath
+  if password is not None:
+    securesystemslib.formats.PASSWORD_SCHEMA.check_match(password)
+    # No additional vetting needed. Decryption will show if it was correct.
 
+  return password
 
 
 
-def import_rsa_privatekey_from_file(filepath, password=None,
-    scheme='rsassa-pss-sha256', prompt=False,
-    storage_backend=None):
-  """
-  <Purpose>
-    Import the PEM file in 'filepath' containing the private key.
 
-    If password is passed use passed password for decryption.
-    If prompt is True use entered password for decryption.
-    If no password is passed and either prompt is False or if the password
-    entered at the prompt is an empty string, omit decryption, treating the
-    key as if it is not encrypted.
-    If password is passed and prompt is True, an error is raised. (See below.)
+def generate_and_write_rsa_keypair(filepath=None, bits=DEFAULT_RSA_KEY_BITS,
+    password=None, prompt=False):
```

Comment:
general comment on `generate_*()` functions: If I've understood correctly calling them with default values now leads to creating unencrypted keys. It's not horrible but I think the expected default would be prompting...

Instead of `prompt` these functions could maybe have `encrypted=True`? By default this would lead to a prompt if password was not given, but you could set `encrypted=False` to create unencrypted keys.

Reply:
Good points, @jku! 

We discussed using an `encrypt` instead of a `prompt` parameter in https://github.com/secure-systems-lab/securesystemslib/pull/124#issuecomment-371621875 pp. but opted for `prompt` to be more explicit about whether user interaction is required. However, I do see how one could equally well argue for being more explicit about encryption. 

Regarding default, I agree that from a security point of view `prompt=True` makes more sense (i.e. secure by default). From a usability point of view I don't like that an API function opens a prompt by default. But the former is probably more important. What do others think? (cc @joshuagl, @mnm678, @adityasaky, @SantiagoTorres,  @trishankatdatadog).

Reply:
> I don't like that an API function opens a prompt by default

Fair point, understood.

I guess you could decide that user has to choose something: leaving both `password` and `prompt` unset would be an error. Then calling the function with default argument values would fail with ValueError instead of prompting unexpectedly.

To create unencrypted keys you would have to `generate_and_write_rsa_keypair(filepath, prompt=False)`: it's not ideal because nothing in that explicitly shows that the result is going to be unencrypted... so it's still not great.

Reply:
Yes, we should encrypt by default, no question about it, and principle of least surprise. Could the API be designed better? Sure, but let's leave that for the future.

Reply:
I think we should only create unencrypted keys if the user explicitly sets this, for example `encrypted=False`, or at the very least we can set `prompt=True` by default.

Reply:
An API function that opens a prompt by default does seem suboptimal, but we can mitigate that somewhat with documentation making clear that this default is only to ensure security by default. I agree with the sentiment that secure by default would be preferable and less surprising.

Reply:
> I guess you could decide that user has to choose something: leaving both password and prompt unset would be an error. Then calling the function with default argument values would fail with ValueError instead of prompting unexpectedly.

Do you mean?:
```
(password=None, prompt=None)  --> error (default)
(password=None, prompt=True)  --> prompt
(password=None, prompt=False) --> plain
(password="..", prompt=None)  --> encrypt
(password="..", prompt=True)  --> error
(password="..", prompt=False) --> encrypt
```
I think we can do that, but I'm not a big fan of raising an exception per default, nor of using 3 values for a boolean.

> To create unencrypted keys you would have to generate_and_write_rsa_keypair(filepath, prompt=False): it's not ideal because nothing in that explicitly shows that the result is going to be unencrypted... so it's still not great.

If we go with above suggestion, I think it makes more sense to use `encrypt` instead of `prompt`. 

> I guess a combo is possible as well: user needs to choose exactly one of password=<value>, prompt=True or encrypted=False -- then each call would be explicit about what it wants. In another language this would be a SecurityType enum with three possible choices but that seems a bit clunky in Python

That  -- i.e.  pass a subclass of `KeySerializationEncryption`, e.g. `BestAvailableEncryption(password)` or `NoEncryption` --  is what [`pyca/cryptography` does](https://cryptography.io/en/latest/hazmat/primitives/asymmetric/serialization.html#serialization-encryption-types). I wouldn't say that it is generally clunky in Python, but it doesn't really fit in with the general style of our interface.

Reply:
I think we have to choose between:
1. default encryption, default prompt
1. no default encryption, no default prompt
1. more sophisticated parameters

Reply:
Yes, you understood my ideas as I meant them. I agree with your comments (raising an exception as default is not ideal, enum style doesn't fit API well)

Reply:
@trishankatdatadog and @joshuagl, you both approved the PR (with no default encryption) but also signaled that you would like to see default encryption. I am unsure how to proceed.

My preferred course of action would be to merge as is, i.e. without default encryption, for the the following (partially interdependent) reasons:
- I'd like to land this asap, because it blocks in-toto 1.0.0.
- I have already adopted the changes from this PR on in-toto (https://github.com/in-toto/in-toto/pull/402) and tuf (will PR later) sides.
- I am against default prompts (see UX).
- I am against more sophisticated parameters (see UX/consistency and *asap*-argument).
- We can reconsider this for securesystemslib v1.0.0. Who knows, we might even remove key generation functionality and limit `securesystemslib` to parsing standard, out-of-band generated key formats, translating them into tuf/in-toto formats, and using them to sign/verify.
- I agree that default encryption is desirable but does not outweigh all these arguments


Can I get 👍 /👎  if others are okay with this?

Reply:
I'm OK with merging as-is and planning to improve in the future by either: 1) removing key generation functionality 2) implementing encryption by default.

Should we file an issue in the 1.0 milestone to capture the discussion here?

Reply:
I wanted to add some comments I shared privately. It would worry me that an unintended side-effect is for the function to block (waiting for stdin), as I assume this would catch many people by surprise, may be difficult to debug (e.g., a CI build hanging, or tox randomy hanging with no stdout/in redirected anywhere).

I'm not entirely sure what's the user story we have in mind, but for a library that is generating keys in a program, it also feels like this is a feature that many people may not want. This is reminiscent of that time we added colorama, and all the consequences that ensured (i.e., very little benefit, and a lot of maintenance burden).

Although I agree that secure by default is a thing, I also wonder if this is adding meaningful security for most of the usecases. I'd expect the following:

- The API function to be noninteractive (hell, we don't even know if this is going to be attached to a terminal)
- A CLI tool that uses this function to set up a prompt (or use this batteries-included setting). 

I don't think at the time of writing this function we have enough context to know how users will use this regularly, so I err on the side of non-interactivity by default.

Reply:
> @trishankatdatadog and @joshuagl, you both approved the PR (with no default encryption) but also signaled that you would like to see default encryption. I am unsure how to proceed.

My bad. I would rather that we encrypt keys on disk rather than our users shoot themselves in the foot. Having said that, TUF/in-toto can take care of that rather than SSLlib. Does this answer your question?